src/mlpack/core/metrics/bleu_impl.hpp

```diff
@@ -0,0 +1,178 @@
+/**
+ * @file core/metrics/bleu_impl.hpp
+ * @author Mrityunjay Tripathi
+ *
+ * Implementation of BLEU class.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_CORE_METRICS_BLEU_IMPL_HPP
+#define MLPACK_CORE_METRICS_BLEU_IMPL_HPP
+
+// In case it hasn't been included.
+#include "bleu.hpp"
+
+namespace mlpack {
+namespace metric {
+
+template <typename ElemType, typename PrecisionType>
+BLEU<ElemType, PrecisionType>::BLEU(const size_t maxOrder) :
+  maxOrder(maxOrder),
+  translationLength(0),
+  referenceLength(0)
+{
+  // Nothing to do here.
+}
+
+template <typename ElemType, typename PrecisionType>
+template <typename WordVector>
+std::map<WordVector, size_t> BLEU<ElemType, PrecisionType>::GetNGrams(
+  const WordVector& segment)
+{
+  std::map<WordVector, size_t> ngramsCount;
+  for (size_t order = 1; order < maxOrder + 1; ++order)
+  {
+    for (size_t i = 0; i < segment.size() - order + 1; ++i)
+    {
+      WordVector seq = WordVector(segment.begin() + i,
+                                  segment.begin() + i + order);
+      ngramsCount[seq]++;
+    }
+  }
+  return ngramsCount;
+}
+
+template <typename ElemType, typename PrecisionType>
+template <typename ReferenceCorpusType, typename TranslationCorpusType>
+ElemType BLEU<ElemType, PrecisionType>::Evaluate(
+  const ReferenceCorpusType& referenceCorpus,
+  const TranslationCorpusType& translationCorpus,
+  const bool smooth)
+{
+  // WordVector is a string container type.
+  // Also, TranslationCorpusType is an array of such containers.
+  typedef typename TranslationCorpusType::value_type WordVector;
+
+  // matchesByOrder: It catches how many times sequence of a particular order
+  // is encountered in both reference corpus and translation corpus.
+  std::vector<size_t> matchesByOrder(maxOrder, 0);
+
+  // possibleMatchesByOrder: It tracks how many possible matches can be in the
+  // translation corpus.
+  std::vector<size_t> possibleMatchesByOrder(maxOrder, 0);
+
+  // referenceLength: It is the sum of minimum length of the paragraph from
+  // various documents.
+  // translationLength: It is the sum of length of each paragraphs.
+  referenceLength = 0, translationLength = 0;
+
+  auto refIt = referenceCorpus.cbegin();
+  auto trIt = translationCorpus.cbegin();
+  for (; refIt != referenceCorpus.cend(), trIt != translationCorpus.cend();
+      ++refIt, ++trIt)
+  {
+    size_t min = std::numeric_limits<size_t>::max();
+    for (auto t : *refIt)
```

Comment:
I think it should be a reference:
```suggestion
    for (auto& t : *refIt)
```